azurerm/resource_arm_image.go

```diff
@@ -0,0 +1,417 @@
+							}, true),
+							DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+						},
+
+						"managed_disk_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+							Optional: true,
+						},
+
+						"blob_uri": {
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+						},
+
+						"caching": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  string(compute.None),
+							ValidateFunc: validation.StringInSlice([]string{
+								string(compute.None),
+								string(compute.ReadOnly),
+								string(compute.ReadWrite),
+							}, true),
+							DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+						},
+
+						"size_gb": {
+							Type:     schema.TypeInt,
+							Computed: true,
+							Optional: true,
+						},
+					},
+				},
+			},
+
+			"data_disk": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+
+						"lun": &schema.Schema{
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+
+						"managed_disk_id": {
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+						},
+
+						"blob_uri": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+
+						"caching": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  "None",
+							ValidateFunc: validation.StringInSlice([]string{
+								string(compute.None),
+								string(compute.ReadOnly),
+								string(compute.ReadWrite),
+							}, true),
+							DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+						},
+
+						"size_gb": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+					},
+				},
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceArmImageCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient)
+	imageClient := client.imageClient
+
+	log.Printf("[INFO] preparing arguments for AzureRM Image creation.")
+
+	name := d.Get("name").(string)
+	location := d.Get("location").(string)
+	resGroup := d.Get("resource_group_name").(string)
+	tags := d.Get("tags").(map[string]interface{})
+	expandedTags := expandTags(tags)
+	properties := compute.ImageProperties{}
+
+	osDisk, err := expandAzureRmImageOsDisk(d)
+	if err != nil {
+		return err
+	}
+
+	dataDisks, err := expandAzureRmImageDataDisks(d)
+	if err != nil {
+		return err
+	}
+
+	storageProfile := compute.ImageStorageProfile{
+		OsDisk:    osDisk,
+		DataDisks: &dataDisks,
+	}
+
+	sourceVM := compute.SubResource{}
+	if v, ok := d.GetOk("source_virtual_machine_id"); ok {
+		vmID := v.(string)
+		sourceVM = compute.SubResource{
+			ID: &vmID,
+		}
+	}
+
+	//either source VM or storage profile can be specified, but not both
+	if (compute.SubResource{}) == sourceVM {
+		//if both sourceVM and storageProfile are empty, return an error
+		if (compute.ImageStorageProfile{}) == storageProfile {
```

Comment:
could we simplify this to `if storageProfile.osDisk == nil && len(storageProfile.dataDisks) == 0`?